app/Domains/Tags/Actions/UpdateTag.php

```diff
@@ -28,30 +28,22 @@ public function authorize(ActionRequest $request): Response
     public function handle(Tag $tag, array $params): Tag
```

Comment:
I am having a hard time understanding the code in `handle`.
Could you please explain it?

Reply:
I think I have made a mistake here, I copied the code handle function from createtag which I should not have done.

Reply:
For updatetag action, in the handle function I need to check whether the updated name represents already existing key or not. If not then new key would be constructed for the updated name to be returned

Reply:
Phase 1

Inputs:  new values in params and the tag model instance
Output: change in the values of the tag being updated
logic/process: with the new params we will check whether it already exists or not. If exists, we will not update otherwise we will construct a new key for the new name and update the tag.

Reply:
Phase 3

Line 1, PHP file has been opened.

Line 3, strict type checking enabled

Line 5, the namespace is App\Domains\Tags\Actions, so to reach this file we have to go to App folder then Domains folder then Tags folder and finally to Actions folder where we will find the file

Line 7, we have injected dependency Tag which is located in App\Domains\Tags\Tag
Line 8, we have injected dependency Response from Illuminate\Auth\Access\Response
Line 9, we have injected dependency Request from Illuminate\Http\Request
Line 10, we have injected dependency ActionRequest from Lorisleiva\Actions\ActionRequest
Line 11, we have injected dependency AsAction from Lorisleiva\Actions\Concerns\AsAction

Line 13, the class name UpdateTag  which is also the file name

Line 15, we instructed the class to inherit the AsActions trait that was injected in line 11

Line 17, we have declared a function called authorize which accepts a HTTP request checking whether user has access or not. It returns the response model.

Line 19, it gets the user making the request

Line 21, the if condition checks if user has general access or not
Line 22, this returns response as true if the condition is true

Line 24, otherwise returns false with the message "You are unauthorised to perform this action"

Line 27, a function called handle is written taking tag model as instance and an array called params. The return type of the function is Tag model instance

Line 30, It is being checked whether the name from the params already exists in the database or not.

Line 32, the condition checks if any similar tag is found or not
Line 33, returns tag without updating anything

Line 36, key for new name is generated
Line 37, tag is updated
Line 38, tag is returned

Line 41, rules function is written whose return type is array

Line 44, id is being validated it being required and it is present in the is column of the tags table
Line 45, name is being validated it being not always required and being string with maximum 255 characters

Line 49, a asController function is written with tag model instance and HTTP request as its input parameter

Line 51, the function is passing the tag model instance and validated request through the handle function and its output is then being returned

Line 54, the jsonResponse function is taking tag model instance and HTTP request as input and the return type is array

Line 57, the function is returning an associative array which contains message as the key and 'Tag updated successfully as the value'

Reply:
GetTag Action

Phase 1

Inputs: id as string of the tag we want to get
Outputs: the desired tag or null if not found
Logic/Process: finding the tag with the given id

Reply:
GetTag Action

Phase 3

Line 1, the php file is opened

Line 3, declaration made that the file is strictly typed

Line 5, namespace is declared which says the file can be found by going to App folder then to Domains folder then Tags and then to Actions and then finally to Queries

Line 7,  dependency Tag is in injected which can be found from App\Domains\Tags\Tag
Line 8, dependency response is injected from Illuminate\Support\Facades\Response
Line 9, dependency ActionRequest is injected from Lorisleiva\Actions\ActionRequest
Line 10, dependency AsAction is injected from Lorisleiva\Actions\Concerns\AsAction

Line 12, class name is written, GetTag which is also the file name

Line 14, AsAction is inherited that we injected in line 10

Line 16, a function called authorize is written which is accepting HTTP request  and the return type is Response

Line 18, the user variable stores the type of user

Line 20, this if statement checks whether the user has general access or not
Line 21, the if statement returns allowed if true

Line 24, otherwise the function returns denied with a message 'You are unauthorized to perform this action'

Line 27, handle function is written which is accepting id as string and the return type is Tag model instance or null

Line 29, the function returns the tag which matches with the given id from the tag database table

Line 32, rules function is made which has no inputs but has output as an array

Line 35, the function validates the id by mentioning it to be required and that the id exists in the id column of the tags table

Line 39, asController function is made which accepts HTTP request

Line 41, returns the output of handle function which takes the id of the requested tag from the request

Line 44, jsonResponse function is written which accepts the tag model instance as the input parameter and the return type is array

Line 47, the function returns an array containing data of tag
Line 48, and the success message

Reply:
GetTags Action

Phase 1

Inputs: first input is the integer which represents the number of items to be shown per page and it is conditional so sometimes may not be given, its default value is 10. It also takes the search term as input
Outputs: the output can be all the tags if the per page value not given or the number of tags to be shown as output will be paginated by the per page value
Logic/Process: the logic is to get all the tags if search term not provided. If provided then based on the search term the tags table will be searched to return any tags if found.

Reply:
GetTags Action

Phase 3

Line 1, php file is opened

Line 3, declaration made that the file is strictly typed

Line 5, namespace is written which tells the destination of the file

Line 7, Tag dependency is injected from App\Domains\Tags\Tag
Line 8, LengthAwarePaginator dependency is injected from Illuminate\Contracts\Pagination\LengthAwarePaginator
Line 9, Collection dependency is injected from Illuminate\Database\Eloquent\Collection
Line 10, Response dependency is injected from Illuminate\Support\Facades\Response
Line 11, ActionRequest dependency is injected from Lorisleiva\Actions\ActionRequest
Line 12, AsAction dependency is injected from Lorisleiva\Actions\Concerns\AsAction

Line 16, the name of the class is written which is also the name of the file

Line 17, AsAction is inherited from where it was injected in line 12

Line 19, a fuction called authorize is written which accepts a HTTP request as input and return type is Response instance

Line 21, user variable stores the data of the user making the request

Line 23, if condition checks whether user has general access or not
Line 24, the condition returns allow if has general access

Line 27, otherwise the function returns deny with a message

Line 30, a handle function is written which accepts an integer value for the number of items to be shown per page with default value 10 and a search term is being taken as input with default value empty string. The return typw is Collection or LengthAwarePaginator

Line 32, Query builder is initiated for Tag

Line 34, the if condition checks whether a search term was given or not
Line 36, the search term is being searched in the name column of tag table
Line 39, query is being returned based on a condition
Line 40, the get query gets all the tags
Line 41, the query data is shown with a limit per page

Line 44, rules function is written whose return type is array

Line 47, the search term should be a string and nullable
Line 48, per_page value should be an integer, nullable and minimum one value should be there

Line 52, asController function accepts a HTTP request

Line 54, search_term variable takes the value as input
Line 55, per_page variable takes value as input

Line 57, the output of the handle function is being returned

Line 60, the jsonResponse function takes tag model instance and HTTP request as input where return type is array

Line 62, message variable stores the count of tags
Line 63, message variable stores found or fetched as string based on search term

Line 66, stores tag model instance as value and data as key in a array
Line 67, a meesage is stored in a message variable